ols/utils/auth_dependency.py

```diff
@@ -0,0 +1,101 @@
+"""Handles authentication in a FastAPI app,.
+
+integrating with Kubernetes SSAR and RBAC for auth checks.
+"""
+
+import json
+import logging
+
+import kubernetes.client
+import kubernetes.utils
+from fastapi import HTTPException, Request
+from urllib3.exceptions import MaxRetryError
+
+from ols.utils import config
+
+logger = logging.getLogger(__name__)
+
+
+async def auth_dependency(request: Request):
+    """Authenticate API requests using Kubernetes SSAR.
+
+    Validates the authorization header and bearer token in the request. If authentication
+    is disabled in the configuration, it's skipped.
+
+    Args:
+        request (Request): The incoming request object from FastAPI.
+
+    Raises:
+        HTTPException: If authentication fails or headers are missing.
+    """
+    if config.dev_config.disable_auth:
+        logger.warn("Auth checks disabled, skipping")
+        return
+    else:
+        # Validate the presence and format of the authorization header
+        authorization_header = request.headers.get("Authorization")
+        if not authorization_header:
+            raise HTTPException(
+                status_code=401, detail="Unauthorized: No auth header found"
+            )
+        # Split the header to extract the token and validate its presence
+        token = _extract_bearer_token(authorization_header)
+        if not token:
+            raise HTTPException(
+                status_code=401,
+                detail="Unauthorized: Bearer token not found or invalid",
+            )
+        # Perform authentication with Kubernetes
+        if not _k8s_auth(token):
+            raise HTTPException(
+                status_code=403, detail="Forbidden: Authentication failed"
+            )
+
+
+def _k8s_auth(api_key) -> bool:
+    try:
+        configuration = kubernetes.client.Configuration()
+        configuration.api_key["authorization"] = api_key
+        configuration.api_key_prefix["authorization"] = "Bearer"
+        configuration.host = config.ols_config.authentication_config.k8s_cluster_api
+        configuration.verify_ssl = True
+
+        if config.ols_config.authentication_config.k8s_ca_cert_path:
+            configuration.ssl_ca_cert = (
+                config.ols_config.authentication_config.k8s_ca_cert_path
+            )
+        if config.ols_config.authentication_config.skip_tls_verification:
+            configuration.verify_ssl = False
+
+        k8s_client = kubernetes.client.ApiClient(configuration)
```

Comment:
instead of having to create a new client w/ the configuration every time when the only thing that is changing is the authorization header, we should be able to create a single client with our own OLS token which has permission to do SubjectAccessReviews(SAR).

That single client can then be called repeatedly, each time performing a SAR, passing the user's access token associated w/ the request and asking k8s if that subject(token) has access to the /ols resource.

Reply:
the New implementation recently pushed, has the centralized client.
but in case of using centralized client, we can make SAR instead of SSAR
for SAR, we make use of user identity
for that i also had to implement fetching user identity feature as a prerequisite.